src/editor/components/StylesColor.js

```diff
@@ -24,22 +25,28 @@ const Color = ( { selector } ) => {
 		themeConfig
 	);
 
-	const onChange = ( newValue, key ) => {
+	/**
+	 * Function to handle the colour palette changes
+	 *
+	 * @param {string} newValue The value of the setting
+	 * @param {string} key      The key of the setting
+	 */
+	const onChange = debounce( ( newValue, key ) => {
```

Comment:
This debounces the `onChange` handler, meaning that the value(s) will only be updated after a delay, which can cause some problems. Ultimately this means we can only really debounce it very slightly, and the original issue where the UI jumps around is still partially present.

What I would suggest is to debounce the values that are passed to the contrast checker as `textColor` and `backgroundColor` props. This would allow the `onChange` to continue to happen immediately, and we can afford to debounce the contrast checker props with a much higher value, if needed, without impacting the UX elsewhere.

This may be easiest with a custom hook, ie:

```js
const useDebouncedValue = ( value, delay ) => {
	const [ debouncedValue, setDebouncedValue ] = useState( value );

	const debouncedSetValue = debounce( ( newValue ) => {
		setDebouncedValue( newValue );
	}, delay );

	useEffect( () => {
		debouncedSetValue( value );
	}, [ value, debouncedSetValue ] );

	return debouncedValue;
};

// usage:
<ContrastChecker
	textColor={ useDebouncedValue(
		varToHex( colorStyles.text, themePalette ),
		500
	) }
	backgroundColor={ useDebouncedValue(
		varToHex( colorStyles.background, themePalette ),
		500
	) }
/>
```

Reply:
Yeah, I think the key difference is the move from debouncing the values (specifically the values passed to the contrast checker) to debouncing the actual `onChange` which results in the values not being updated immediately.

What you had looks good, but there are a couple of improvements to be made:
1. [You can't choose what goes into the dependency array](https://react.dev/reference/react/useEffect#specifying-reactive-dependencies). A fix here would be to pass the values when calling the `debouncedUpdateColors` function, thus making them dependencies of the effect, and update the function to use the values you passed in. Currently, the values being accessed in the function may not be the same as the values that triggered the effect to run.
2. My suggestion also fell foul of this, but we should [remove the function dependency from the effect](https://react.dev/reference/react/useEffect#removing-unnecessary-function-dependencies). Note - moving the function into the effect means that you are making the values dependencies of the effect so no need to pass them into the function 🙌 

So this:
```js
const debouncedUpdateColors = debounce( () => {
		setTextColour( colorStyles.text );
		setBackgroundColour( colorStyles.background );
	}, 150 );

useEffect( () => {
	debouncedUpdateColors();
	return () => {
		debouncedUpdateColors.cancel();
	};
}, [ colorStyles.text, colorStyles.background, debouncedUpdateColors ] );
```

Would become this:
```js
useEffect( () => {
	const debouncedUpdateColors = debounce( () => {
		setTextColour( colorStyles.text );
		setBackgroundColour( colorStyles.background );
	}, 150 );
	debouncedUpdateColors();
	return () => {
		debouncedUpdateColors.cancel();
	};
}, [ colorStyles.text, colorStyles.background, setTextColour, setBackgroundColour ] );
```

The `cancel` is a great shout, nice one!

I think extracting to a custom hook makes sense for this, it keeps things cleaner here and may be useful elsewhere. So the ideal result will be a combination of both of our approaches and my suggestion above. I haven't tested this but it should work 🤞

Reply:
Applying this approach works nicely, a quick demo can be [seen here](http://bigbite.im/v/e9xpAW). Thanks for the advice!

There is still a slight jump in the UI if holding a colour for an extended time within a palette, however it does not hinder the colour selection or cause an infinite glitch as it did previously.

Reply:
Cheers, I think it's in a good enough place to proceed. Happy to keep it in this component for now rather than a custom hook, we can extract in future if we encounter more need for debouncing values